### PR TITLE
feat(github): support fine-grained token

### LINKED ISSUE #18459: Support github fine-grained token
### What would you like Renovate to be able to do?

Github recently introduced fine-grained tokens - https://github.blog/2022-10-18-introducing-fine-grained-personal-access-tokens-for-github/

Will be nice if renovate can support them.

Here is official github discussion related to that feature https://github.com/community/community/discussions/36441

### If you have any ideas on how this should be implemented, please tell us here.

Looks like new method needs to be added. Currently, if used with fine-grained token, github will return following error:

```
 "response": {
      "statusCode": 401,
      "statusMessage": "Unauthorized",
      "body": {
        "message": "Personal access tokens with fine grained access do not support the GraphQL API",
        "documentation_url": "https://docs.github.com/graphql/guides/forming-calls-with-graphql#authenticating-with-graphql"
      },
...
```

### Is this a feature you are interested in implementing yourself?

No

### PR DESCRIPTION
**Pull request by frantic1048**
## Changes

This PR removes limitation on usage of GitHub fine-grained token.

## Context

Resolve issue: https://github.com/renovatebot/renovate/issues/18459

Depends on:
- https://github.com/renovatebot/renovate/pull/23869

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [x] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/workers/global/config/parse/env.spec.ts [hunk 1]
```diff
@@ -161,13 +161,20 @@ describe('workers/global/config/parse/env', () => {
       });
     });
 
-    it('does not support GitHub fine-grained PATs', () => {
+    it('supports GitHub fine-grained PATs', () => {
       const envParam: NodeJS.ProcessEnv = {
         GITHUB_COM_TOKEN: 'github_pat_XXXXXX',
         RENOVATE_TOKEN: 'a github.com token',
       };
       expect(env.getConfig(envParam)).toMatchSnapshot({
```
Comment (viceice) [inline lib/workers/global/config/parse/env.spec.ts:169]:
```suggestion
       expect(env.getConfig(envParam)).toMatchObject({
```

File: lib/modules/platform/github/index.ts [hunk 2]
```diff
@@ -133,11 +133,6 @@ export async function initPlatform({
   if (!token) {
     throw new Error('Init: You must configure a GitHub token');
   }
-  if (token.startsWith('github_pat_')) {
```
[the remark this answers is not in the document]
Comment (frantic1048) [reply lib/modules/platform/github/index.ts:136]: I added checks when fine-grained token and GHE are both met at [`cd3cefd` (#23810)](https://github.com/renovatebot/renovate/pull/23810/commits/cd3cefda1d37015e5de805160db24fb952535252)

File: lib/modules/platform/github/index.ts [hunk 3]
```diff
@@ -144,6 +145,21 @@ export async function initPlatform({
   }
 
   await detectGhe(token);
+  /**
+   * GHE requires version >=3.10 to support fine-grained access tokens
+   * https://docs.github.com/en/enterprise-server@3.10/admin/release-notes#authentication
+   */
+  if (isGithubFineGrainedPersonalAccessToken(token) && platformConfig.isGhe) {
+    if (!platformConfig.gheVersion) {
```
Comment (frantic1048) [inline lib/modules/platform/github/index.ts:153]: I'm unsure if this check is necessary, or if these two conditions should be combined into one with a unified error message.

I noticed a test case in `lib/modules/platform/github/index.spec.ts`: ___should support custom endpoint without version___, which adds this check when no version is detected from GHE.

However, I am unsure in which scenario GHE omits the version from the response, or if this case should prevent Renovate from proceeding.

File: lib/modules/platform/github/index.spec.ts [hunk 4]
```diff
@@ -57,14 +57,50 @@ describe('modules/platform/github/index', () => {
...
+        .get('/user')
+        .reply(200, { login: 'renovate-bot' })
+        .get('/user/emails')
+        .reply(200, [{ email: 'user@domain.com' }]);
+      expect(
+        await github.initPlatform({
+          endpoint: 'https://ghe.renovatebot.com',
+          token: 'github_pat_XXXXXX',
+        })
+      ).toMatchSnapshot();
```
Comment (viceice) [inline lib/modules/platform/github/index.spec.ts:101]: use `toEqual`

File: lib/modules/platform/github/index.ts [hunk 5]
```diff
@@ -38,6 +38,7 @@ import { regEx } from '../../../util/regex';
 import { sanitize } from '../../../util/sanitize';
 import { fromBase64, looseEquals } from '../../../util/string';
 import { ensureTrailingSlash } from '../../../util/url';
+import { isGithubFineGrainedPersonalAccessToken } from '../../manager/composer/utils';
```
Comment (viceice) [inline lib/modules/platform/github/index.ts:41]: Move that function to utils in separate PR before merging this PR

File: lib/modules/platform/github/index.ts (hunk 3 above)
Comment (viceice) [reply lib/modules/platform/github/index.ts:153]: combine to one error, it should fail for unknown and old versions.

missing version header typically means a very old ghes.

File: lib/workers/global/config/parse/env.spec.ts [hunk 6]
```diff
@@ -161,13 +161,20 @@ describe('workers/global/config/parse/env', () => {
...
     });
 
-    it('does not support GitHub fine-grained PATs', () => {
+    it('supports GitHub fine-grained PATs', () => {
       const envParam: NodeJS.ProcessEnv = {
         GITHUB_COM_TOKEN: 'github_pat_XXXXXX',
         RENOVATE_TOKEN: 'a github.com token',
       };
-      expect(env.getConfig(envParam)).toMatchSnapshot({
+      expect(env.getConfig(envParam)).toMatchObject({
```
Comment (viceice) [inline lib/workers/global/config/parse/env.spec.ts:169]:
```suggestion
      expect(env.getConfig(envParam)). toEqual({
```

File: lib/modules/platform/github/index.ts (hunk 5 above)
Comment (frantic1048) [reply lib/modules/platform/github/index.ts:41]: That function `isGithubFineGrainedPersonalAccessToken()` along with similar util functions are already in Renovate repo:

https://github.com/renovatebot/renovate/blob/666390468c3969bdcc2941147a0b5b0908110eec/lib/modules/manager/composer/utils.ts#L121-L131

Should I move them to another module, and where is the proper place to put them?

**Comment by frantic1048** [comment]
I combined the unknown and old GHE version checks into one, PTAL 🙏

File: lib/modules/platform/github/index.ts (hunk 5 above)
Comment (viceice) [reply lib/modules/platform/github/index.ts:41]: https://github.com/renovatebot/renovate/blob/main/lib/util/check-token.ts

I think that is a good place

**Review by viceice — COMMENTED** [review]
see comment

File: lib/modules/platform/github/index.ts (hunk 5 above)
Comment (frantic1048) [reply lib/modules/platform/github/index.ts:41]: I opened a seperate PR for this change: https://github.com/renovatebot/renovate/pull/23869

Once above PR is merged, I have to rebase this PR to use proper import path. If no further change is required for this PR, I may rebase this PR as soon as possible after https://github.com/renovatebot/renovate/pull/23869 is merged. <- Is it ok to do so?

Comment (viceice) [reply lib/modules/platform/github/index.ts:41]: no need to rebase, just merge main and fix conflicts. we'll squash merge anyways

Comment (frantic1048) [reply lib/modules/platform/github/index.ts]: ok, I have merged main and updated the import path for `isGithubFineGrainedPersonalAccessToken()`. The PR looks up to date now.

**Review by viceice — APPROVED** [review]
